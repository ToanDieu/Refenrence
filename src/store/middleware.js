import { matchPath } from "react-router-dom";
import { CHANGE_PAGE, NOTIFY_BUTTON_CLICK } from "../constants/actionTypes";

const changeToBasePage = id => ({
  type: CHANGE_PAGE,
  payload: {
    pageType: "base",
    detail: { id }
  }
});

export const trackPageNavigate = ({
  dispatch
  // getState
}) => next => action => {
  if (action.type === "@@router/LOCATION_CHANGE") {
    console.log("trackViews__action: ", action);
    const pathName = action.payload.location.pathname;

    const isEnteringBaseCases = matchPath(pathName, {
      path: "/bases/:baseID/cases"
    });

    const isEnteringBaseDetail = matchPath(pathName, {
      path: "/bases/:baseID"
    });

    // console.log("isEnteringBaseDetail", isEnteringBaseDetail, action);

    const isEnteringBaseWorkflow = matchPath(pathName, {
      path: "/bases/:baseID/workflows/*/actions"
    });

    const isEnteringNotification = matchPath(pathName, {
      path: "/notifications"
    });

    if (isEnteringNotification) {
      dispatch({
        type: NOTIFY_BUTTON_CLICK
      });
    }

    if (isEnteringBaseCases) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseCases.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }

    if (isEnteringBaseDetail) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseDetail.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }

    if (isEnteringBaseWorkflow) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseWorkflow.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }
  }

  return next(action);
};
