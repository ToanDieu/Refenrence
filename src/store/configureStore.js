import { createStore, compose, applyMiddleware } from "redux";
import reduxImmutableStateInvariant from "redux-immutable-state-invariant";
import thunk from "redux-thunk";
import { routerMiddleware } from "connected-react-router";

import history from "@/utils/history";
import createReducer from "../reducers";
import { trackPageNavigate } from "./middleware";

function configureStoreProd(initialState) {
  const reactRouterMiddleware = routerMiddleware(history);
  const middlewares = [thunk, reactRouterMiddleware, trackPageNavigate];

  const store = createStore(
    createReducer(),
    initialState,
    compose(applyMiddleware(...middlewares))
  );

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  return store;
}

function configureStoreDev(initialState) {
  const reactRouterMiddleware = routerMiddleware(history);
  const middlewares = [
    reduxImmutableStateInvariant(),
    thunk,
    reactRouterMiddleware,
    trackPageNavigate
  ];

  /* eslint-disable no-underscore-dangle */
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; // add support for Redux dev tools
  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(applyMiddleware(...middlewares))
  );

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept("../reducers", () => {
      const nextReducer = require("../reducers").default; // eslint-disable-line global-require
      store.replaceReducer(nextReducer);
    });
  }

  return store;
}

const store =
  process.env.NODE_ENV === "production"
    ? configureStoreProd()
    : configureStoreDev();

export { history, store };
