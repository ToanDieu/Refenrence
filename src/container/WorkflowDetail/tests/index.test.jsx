import React from "react";
import { render } from "react-testing-library";
import { Provider } from "react-redux";
import { initialize, addTranslationForLanguage } from "react-localize-redux";
import { Notification } from "element-react";

import languages from "@/i18n/languages.json";
import english from "@/i18n/en.json";
import { store as appStore } from "@/store/configureStore";
import { WorkflowDetail } from "../index";

describe("<WorkflowDetail />", () => {
  let store;

  beforeAll(() => {
    store = appStore;
    const defaultLanguage = "en";
    store.dispatch(initialize(languages, { defaultLanguage }));
    store.dispatch(addTranslationForLanguage(english, "en"));
  });

  it("should render the loading indicator when its loading", () => {
    const { container } = render(
      <Provider store={store}>
        <WorkflowDetail loading />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render the create workflow form & button when there is no workflow", () => {
    const { container } = render(
      <Provider store={store}>
        <WorkflowDetail loading={false} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render the add action button when there is no actions yet", () => {
    const workflow = { id: 1, name: "Test workflow", initialActionID: 1 };
    const actions = [];
    const translate = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <WorkflowDetail
          loading={false}
          workflow={workflow}
          actions={actions}
          translate={translate}
        />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render the workflow graph within actions/triggers", () => {
    const workflow = { id: 1, name: "Test workflow", initialActionID: 1 };
    const actions = [
      { id: 1, acceptActionID: 2, rejectActionID: 3, actionType: "Branch" },
      { id: 2, acceptActionID: undefined, rejectActionID: undefined },
      { id: 3, acceptActionID: undefined, rejectActionID: undefined }
    ];
    const { container } = render(
      <Provider store={store}>
        <WorkflowDetail loading={false} workflow={workflow} actions={actions} />
      </Provider>
    );
    expect(container.firstChild.lastChild).toMatchSnapshot();
  });

  it("should fetch data on mount", () => {
    const spy = jest.fn();
    render(
      <Provider store={store}>
        <WorkflowDetail baseID={1} fetchData={spy} />
      </Provider>
    );
    expect(spy).toHaveBeenCalled();
  });

  it("should show error notification when re-render within error", () => {
    const spy = jest.spyOn(Notification, "error");
    render(
      <Provider store={store}>
        <WorkflowDetail error={{ message: "Error" }} />
      </Provider>
    );
    expect(spy).toHaveBeenCalled();
  });
});
