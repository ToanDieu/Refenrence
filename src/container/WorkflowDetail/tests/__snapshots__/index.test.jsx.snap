// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WorkflowDetail /> should render the add action button when there is no actions yet 1`] = `
.emotion-4 {
  padding: 30px 30px 100px;
  min-height: calc(100vh - 156px);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 60px;
}

.emotion-0 {
  font-size: 22px;
  margin-right: 10px;
  text-transform: uppercase;
}

.emotion-3 {
  text-align: center;
}

<article
  class="emotion-4"
>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
    >
      <span
        class="emotion-0"
      >
        Test workflow
      </span>
      <span />
    </div>
    <div />
  </div>
  <div
    class="emotion-3"
  />
</article>
`;

exports[`<WorkflowDetail /> should render the create workflow form & button when there is no workflow 1`] = `
.emotion-1 {
  padding: 30px 30px 100px;
  min-height: calc(100vh - 156px);
}

.emotion-0 {
  text-align: center;
}

.emotion-0 p {
  margin: 12px;
}

<article
  class="emotion-1"
>
  <div>
    <div
      class="emotion-0"
    >
      <p>
        This base has no workflow yet
      </p>
    </div>
    <div>
      <div
        class="el-dialog__wrapper"
        style="z-index: 1013; display: none;"
      >
        <div
          class="el-dialog el-dialog--tiny"
          style="top: 15%;"
        >
          <div
            class="el-dialog__header"
          >
            <span
              class="el-dialog__title"
            >
              Create workflow
            </span>
            <button
              class="el-dialog__headerbtn"
              type="button"
            >
              <i
                class="el-dialog__close el-icon el-icon-close"
              />
            </button>
          </div>
          <div
            class="el-dialog__body"
          >
            <form
              class="el-form el-form--label-top"
            >
              <div
                class="el-form-item is-required"
              >
                <label
                  class="el-form-item__label"
                >
                  Workflow name
                </label>
                <div
                  class="el-form-item__content"
                >
                  <div
                    class="el-input"
                  >
                    <input
                      autocomplete="off"
                      class="el-input__inner"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="el-form-item"
              >
                <label
                  class="el-form-item__label"
                >
                  Description
                </label>
                <div
                  class="el-form-item__content"
                >
                  <div
                    class="el-input"
                  >
                    <input
                      autocomplete="off"
                      class="el-input__inner"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </form>
          </div>
          <div
            class="el-dialog__footer dialog-footer"
          >
            <button
              class="el-button el-button--primary"
              type="submit"
            >
              <span>
                Create
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="v-modal"
        style="z-index: 1012; display: none;"
      />
    </div>
  </div>
</article>
`;

exports[`<WorkflowDetail /> should render the loading indicator when its loading 1`] = `
.emotion-0 {
  padding: 30px 30px 100px;
  min-height: calc(100vh - 156px);
}

<article
  class="emotion-0"
>
  <div
    class="spinner-wrapper"
  >
    <div
      class="spinner__dots spinner--row"
    >
      <div
        class="bounce1"
      />
      <div
        class="bounce2"
      />
      <div
        class="bounce3"
      />
    </div>
  </div>
</article>
`;

exports[`<WorkflowDetail /> should render the workflow graph within actions/triggers 1`] = `
.emotion-7 ul {
  margin: 0;
  padding: 0;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
}

.emotion-7 ul li {
  position: relative;
}

.emotion-7 ul ul {
  padding-top: 30px;
}

.emotion-7 ul ul::before {
  content: "";
  position: absolute;
  top: 0;
  left: calc(50% - 1px);
  border-left: 2px solid #1E5A6E;
  height: 30px;
}

.emotion-7 ul ul li {
  padding-top: 30px;
}

.emotion-7 ul ul li::before,
.emotion-7 ul ul li::after {
  content: "";
  position: absolute;
  top: 0;
  right: calc(50% - 1px);
  border-top: 2px solid #1E5A6E;
  width: calc(50% + 1px);
  height: 30px;
}

.emotion-7 ul ul li::after {
  right: auto;
  left: calc(50% - 1px);
  border-left: 2px solid #1E5A6E;
}

.emotion-7 ul ul li:first-of-type::before,
.emotion-7 ul ul li:last-of-type::after {
  display: none;
}

.emotion-7 ul ul li:last-of-type::before {
  border-right: 2px solid #1E5A6E;
}

.emotion-7 ul ul li:only-of-type::before {
  display: block;
  border: none;
  left: calc(50% - 1px);
  border-left: 2px solid #1E5A6E;
}

.emotion-0 {
  line-height: 1.45;
  padding-top: 8px;
  white-space: pre-wrap;
}

.emotion-1 {
  margin: 0 auto;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  border-radius: 4px;
  padding: 5px 10px;
  text-transform: uppercase;
  background-color: #26C022;
}

.emotion-3 {
  margin: 0 auto;
  padding: 0px 24px;
}

.emotion-2 {
  margin: 0 auto;
  padding: 16px 0px;
  width: 120px;
  border-radius: 4px;
  text-align: center;
  color: #7f7f7f;
}

.emotion-4 {
  margin: 0 auto;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  border-radius: 4px;
  padding: 5px 10px;
  text-transform: uppercase;
  background-color: #E81919;
}

<div
  class="emotion-7"
>
  <ul>
    <li>
      <div
        css="[object Object]"
      >
        <div
          css="[object Object]"
        >
          <div
            css="[object Object]"
          >
            <div>
              <div
                css="[object Object]"
              >
                <span
                  class="icon-question"
                />
                <span
                  css="[object Object]"
                >
                   
                  Is the following condition(s) matched?
                </span>
              </div>
              <div
                role="presentation"
              >
                <div
                  class="p-workflow-list"
                  css="[object Object]"
                >
                  <div>
                    <div
                      role="presentation"
                    >
                      <div
                        css="[object Object]"
                      >
                        <span
                          class="icon-edit"
                        />
                        <span
                          css="[object Object]"
                        >
                           
                          Edit
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="emotion-0"
                >
                  <span />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <ul>
        <li>
          <div
            class="emotion-1"
          >
            yes
          </div>
          <ul>
            <li>
              <div
                css="[object Object]"
              >
                <div
                  css="[object Object]"
                >
                  <div
                    css="[object Object]"
                  >
                    <div />
                  </div>
                </div>
              </div>
              <ul>
                <li>
                  <div
                    class="emotion-3"
                  >
                    <div
                      class="emotion-2"
                    >
                      End
                    </div>
                  </div>
                </li>
              </ul>
            </li>
          </ul>
        </li>
        <li>
          <div
            class="emotion-4"
          >
            no
          </div>
          <ul>
            <li>
              <div
                css="[object Object]"
              >
                <div
                  css="[object Object]"
                >
                  <div
                    css="[object Object]"
                  >
                    <div />
                  </div>
                </div>
              </div>
              <ul>
                <li>
                  <div
                    class="emotion-3"
                  >
                    <div
                      class="emotion-2"
                    >
                      End
                    </div>
                  </div>
                </li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
